### ISSUE TITLE
DRY refactor — extract repeated notebook patterns to `_common.py`

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/SARS-CoV-2_spike_multidms#40]
# DRY refactor — extract repeated notebook patterns to `_common.py`

## Summary

Several code patterns are duplicated across multiple notebooks: `deep_merge` exists in both the Snakefile and `_common.py`, the `all_muts_known` closure is defined identically in 3 files, `fit_params` dict construction is repeated 7 times, and the subsample logic in `spike_01` uses a hardcoded `random_state=0` that ignores the config seed. Extracting these to `_common.py` reduces maintenance burden and ensures consistent behavior.

## Background

`notebooks/_common.py` already serves as the shared utility module (containing `deep_merge`, `load_config`, `build_phenotype_fxn_dict`, `combine_replicate_muts`). The patterns below should join it.

### Duplicated patterns

1. **`deep_merge`** — Defined identically in `workflow/Snakefile` (line 7) and `notebooks/_common.py` (line 10). The Snakefile version should import from `_common.py` or the function should live in a shared location.

2. **`all_muts_known` closure** — Defined identically in 3 files:
   - `spike_05_cross_validation.py` line 173
   - `spike_08_linear_comparison.py` line 212
   - `sim_08_cross_validation.py` line 143

   Each defines:
   ```python
   def all_muts_known(subs, km=_known_muts):
       return all(s in km for s in subs.split())
   ```

3. **`fit_params` dict construction** — 7 notebooks build nearly identical fitting parameter dicts from config values:
   - `spike_03`, `spike_05`, `spike_07`, `spike_08`, `spike_10` (spike config)
   - `sim_02`, `sim_08` (simulation config)

   Each constructs a dict with keys like `lasso_shifts`, `epistatic_model`, `output_activation`, `n_epochs`, `learning_rate`, etc., pulled from the same config paths.

4. **Subsample logic** — `spike_01` line 212 uses `random_state=0` hardcoded instead of reading from `spike.get("seed")` or the config's seed value:
   ```python
   lambda x: x.sample(frac=subsample_frac, random_state=0)
   ```

5. **Train/test split** — `spike_05` and `spike_08` both implement the same train/test split logic for cross-validation (splitting func_scores by condition, sampling a fraction for test, using the remainder for training).

### 1. Deduplicate `deep_merge`

**Files:** `workflow/Snakefile`, `notebooks/_common.py` (existing)

The Snakefile can't easily import from `notebooks/_common.py` due to Snakemake's module resolution. Options:
- **Option A:** Keep both copies but add a comment in each pointing to the other as the canonical source.
- **Option B:** Move `deep_merge` to a standalone `workflow/scripts/utils.py` and import from both locations.
- **Recommended:** Option A — the duplication is small (8 lines) and the Snakefile's Python environment is constrained.

### 2. Extract `all_muts_known` to `_common.py`

**File:** `notebooks/_common.py` (existing)

```python
def all_muts_known(substitutions, known_muts):
    """Check if all mutations in a substitution string are in the known set.

    Parameters
    ----------
    substitutions : str
        Space-separated mutation strings.
    known_muts : set
        Set of known mutation strings.

    Returns
    -------
    bool
    """
    return all(s in known_muts for s in substitutions.split())
```

Update callers in `spike_05`, `spike_08`, and `sim_08` to import and use:
```python
from notebooks._common import all_muts_known
_mask = _test_df["aa_substitutions"].apply(lambda s: all_muts_known(s, _known_muts))
```

### 3. Extract `build_fit_params` to `_common.py`

**File:** `notebooks/_common.py` (existing)

```python
def build_fit_params(config_section, dataset):
    """Build a standard fit_params dict from a config section.

    Parameters
    ----------
    config_section : dict
        The ``fitting`` subsection of the pipeline config
        (e.g., ``config["spike"]["fitting"]``).
    dataset : multidms.Data
        The dataset to fit.

    Returns
    -------
    dict
        Ready to pass to ``multidms.Model()`` or ``fit_models()``.
    """
    return {
        "dataset": dataset,
        "lasso_shifts": config_section["lasso_shifts"],
        "epistatic_model": config_section.get("epistatic_model", "Sigmoid"),
        "output_activation": config_section.get("output_activation", "Identity"),
        "n_epochs": config_section["n_epochs"],
        "learning_rate": config_section.get("learning_rate", 5e-4),
        # ... additional standard keys
    }
```

The exact keys will need to be determined by inspecting the union of all 7 construction sites. Some notebooks add extra keys (e.g., `spike_07` for single-condition fits) — those can extend the base dict.

### 4. Fix subsample `random_state` to use config seed

**File:** `notebooks/spike/spike_01_data_loading.py` (existing)

```python
# Change from:
lambda x: x.sample(frac=subsample_frac, random_state=0)

# To:
seed = spike.get("seed", 0)
lambda x: x.sample(frac=subsample_frac, random_state=seed)
```

### 5. Extract train/test split helper

**File:** `notebooks/_common.py` (existing)

Extract the shared train/test split logic from `spike_05` and `spike_08` into a helper that takes func_scores, test_frac, and seed, and returns (train_dfs, test_dfs) split by condition.

## Testing Strategy

- Run `pixi run test` end-to-end to confirm all notebooks still execute
- Verify that extracted functions produce identical outputs to the inline versions by comparing pipeline results before and after
- Check that `spike_01` with `subsample_frac` set in config uses the config seed, not hardcoded 0

## Success Criteria

- [ ] `all_muts_known` defined once in `_common.py`, imported by 3 notebooks
- [ ] `build_fit_params` defined once in `_common.py`, used by 7 notebooks
- [ ] Subsample logic uses config seed instead of hardcoded `random_state=0`
- [ ] Train/test split extracted to shared helper
- [ ] `deep_merge` duplication documented or resolved
- [ ] `pixi run test` passes
